### DISCUSSION TITLE
Sporingsidentifikator for sammenstilling av logger i ettertid

### DISCUSSION THREAD

**Discussion by eirikbroen** [NorskHelsenett/Tillitsrammeverk#27] [General]
Vi har tidligere diskutert behovet for en unik identifikator som følger en forespørsel om innsyn fra den konsumerende part til den utleverende part. Behovet for en slik unik identifikator er blant annet.

- Grunnlag for å sammenstille loggdata. For å kunne gjennomføre loggkontroll på tvers av virksomheter er det behov for å gjøre det enkelt å koble sammen en forespørsel med en utlevering. Innføring av en unik identifikator og krav om logging av denne verdien i alle ledd av en kallkjede gir oss et godt utgangspunkt for å analyse både i forbindelse med kontroll og ved feilsituasjoner.
- Aggregering av logginnslag i Helsenorge. Vi bør gjøre det mulig å slå sammen loggdata fra både konsumerende og utleverende part slik at det fremkommer som ett enkelt innslag i innsynslogg for innbygger.

Det er ønskelig at en slik unik identifikator "oppstår" så tidlig i kallkjeden som mulig. Det er også viktig at identifikator er unik på tvers av system i sektor, det kan for eksempel gjøres ved å benytte prefiks tilhørende HelseID klient. 

Eksempel på hvordan identifikator (her kalt ID) kan opptre gjennom et forespørsel om visning av dokument vi Kjernejournal Portal. Eksempelet er forenklet og tar bare med et subsett av de faktiske interaksjonene mellom partene.

```mermaid
sequenceDiagram
participant c as Konsumerende part
participant h as HelseID
participant kj as Kjernejournal Portal
participant x as XDS
participant u as Utleverende part
c->>c: Generer identifikator
c->>c: Bygg autorisasjonsdetaljer med identifikator (ID)
c->>h: POST /authorize (body >> authorization_details {... "ID": "someval"})
h-->>c: OK
c->>h: POST /token 
h-->>c: Access Token (claim: authorization_details {... "ID": "someval"})
h->>h: Logg (token utlevert + ID)
c->>c: Logg (token mottatt + ID)
c->>kj: Innlogget bruker med aktivt Access token
kj->>x: Hent dokument (Access token fra konsumerende part)
kj->>kj: Logg (token benyttet + ID)
x->>x: Logg (token mottat + ID)
x->>u: Hent dokument (access token)
u->>u: Autoriser utlevering basert på token
u->>u: Logg beslutning og utlevering + ID
u-->>x: 200 Dokument
x-->>kj: 200 Dokument
kj-->>c: Vis dokument 
``` 

Vil gjerne at vi diskuterer 

- Er vi enige om behovet for en slik identifikator?
- Hvor skal den oppstå?
- Hva er kobling til logg?

Også interessant å komme inn på hvem som er ansvarlig for å logge annen data vi utveksler som del av tillitsmodellen.

**Comment by stvgr** [comment]
Stavanger tenker at vi i kommunen ikke trenger denne for vår del, vi vil klare å finne rette logg element basert på tidsstempel.
Men det er nok et viktig element hvis loggene skal sammenstilles og presenteres for innbygger.

**Comment by richardhus** [comment]
Støtter @eirikbroen i behovet for en slik identifikator, ref. også annen diskusjon startet av @micer2004 tidligere i dag.

Også enig i at det bør avklares ansvar for logging av annen data som utveksles som del av tillitsmodellen. Blant annet tenker jeg at det er konsumentens ansvar å logge at man avgir informasjon om grunnlaget for tilgangen (behandlerrelasjon++) til HelseID og/eller andre som man avgir dette til. Om eventuell nasjonal tjenesteleverandør og kilder dette overføres til vil være selvstendig dataansvarlige for videre behandling av disse opplysningene (at de logger dem og eventuelt viser dem til innbygger fra sine tilgangslogger er jo også en behandling) er viktig å fastslå i avtaler for tillitsrammeverket.

**Comment by palcarlsen** [comment]
Kan vi ikke benytte den "halvveis standarden" "X-Correlation-Id" som legges i http header? 
Da kan fagsystem opprette en id (det bør benyttes GUID for å være sikker på at denne faktisk blir unik) på den måten som passer inn i organisasjonen/systemet den enkelte aktør sitter på. Så er det en smal sak å legge inn denne verdien i HTTP headeren når man kaller en ekstern aktør. Mottager kan plukke med seg http headeren og inkludere den i egen logging, igjen på en måte som passer med hvordan applikasjonen fungerer.

**Reply by richardhus** [reply]
@palcarlsen: Den standarden er nok primært ment for å korrelere alt som hører til samme tekniske request/response-kjede, mens for å kunne knytte sammen alle request/response-kjeder som er trigget innenfor den samme brukersesjonen for helsepersonellet (som gjelder tilgangen til den spesifikke pasienten) krever en ny type ID som er det vi diskuterer her. Med denne kan vi binde sammen logger mellom forhåpentligvis alle aktører, og så kan den tekniske korreleringsnøkkelen du nevner supplere dette, slik at man også kan følge enkelt request/response-kjeder når man trenger det. Så begge deler er nok aktuelt.
